File: CallsAndSMS/app/src/main/java/nisrulz/github/sample/callsandsms/SMSReceiver.java
package nisrulz.github.sample.callsandsms;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.widget.Toast;

public class SMSReceiver extends BroadcastReceiver {

  @Override
  public void onReceive(Context context, Intent intent) {
    Bundle bundle = intent.getExtras();
    if (bundle != null) {
      Object[] pdus = (Object[]) bundle.get("pdus");
      String format = bundle.getString("format");

      final SmsMessage[] messages = new SmsMessage[pdus.length];
      for (int i = 0; i < pdus.length; i++) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
          messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i], format);
        }
        else {
          messages[i] = SmsMessage.createFromPdu((byte[]) pdus[i]);
        }
        String senderPhoneNo = messages[i].getDisplayOriginatingAddress();
        Toast.makeText(context,
            "Message " + messages[0].getMessageBody() + ", from " + senderPhoneNo,
            Toast.LENGTH_SHORT).show();
      }
    }
  }
}

File: CallsAndSMS/app/src/main/java/github/nisrulz/example/callsandsms/SMSReceiver.kt
package github.nisrulz.example.callsandsms

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.telephony.SmsMessage
import android.util.Log

class SMSReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        intent.extras?.let { bundle ->
            val pdus = bundle["pdus"] as Array<*>?
            val format = bundle.getString("format")

            pdus?.apply {
                val messages = arrayOfNulls<SmsMessage>(size)
                for (index in indices) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        messages[index] = SmsMessage.createFromPdu(pdus[index] as ByteArray, format)
                    } else {
                        messages[index] = SmsMessage.createFromPdu(pdus[index] as ByteArray)
                    }
                    val senderPhoneNo = messages[index]?.displayOriginatingAddress
                    Log.d(
                        "SMSReceiver",
                        "Message " + messages[0]?.messageBody + ", from " + senderPhoneNo,
                    )
                }
            }
        }
    }
}